### ISSUE TITLE
About the memory leak & shape disappearance

### ISSUE THREAD

**Issue by chrisj0192** [pubpub-zz/ppInk#75]
Hello friend! I hope you are well. 
Recently I was testing new features of 1.5 RC 2 ppink and it's amazing thanks for that!
But I have a little issue with the magnetic effect(I've also noticed this issue in other versions), I've noticed that every time I draw curved lines or write letters with the pen, a little magnetic effect appears and it increases more when I draw or write more things on the screen. I set the magnetic radius to zero in the menu but that effect persists and I was wondering if there is a method to reduce that effect, because when I draw a lot of things on the screen a little delay appears and extra lines appears (lines that I don't draw) to due that magnetic effect.
Another thing that I have noticed is that when I draw a lot of figures or things on my screen with the pen, my screen crashes, I don't know what the exact reason is. By the way, my laptop has a good processor (core i7 2.5Ghz, 6th generation) and 8GB ram.

Thanks in advance for your help.

**Comment by pubpub-zz** [comment]
You can switch off the magnetic effect clicking on the icon in the toolbar. When off is it better? You can try also to press shift key when clicking  to only activate a limited magnetic effect. else can you you make some snapshots or videos to illustrate your problem.

About your issue with crash, can you send some logs.

**Comment by chrisj0192** [comment]
Hello friend, yes I tried to disable the magnetic effect by clicking the icon in the toolbar, pressing the shift key and typing 0.0 in the magnetic radius text field but some slight movement persists. Hmm I was wondering if it would be a default movement of lines or strokes to make them sharpness (I'm not sure if this is the right word), but If I draw a lot of things on my screen, that movement of lines becomes stronger and it seems that some extra lines in my stroke appear.

This is a short video about extra lines appears when I draw.
[https://youtu.be/G7oZtCgSsWM](url)

**Comment by pubpub-zz** [comment]
negative value keep track of the magnetic value but is a disabled value. Any way when you are using the hand free drawing, magnetic has no effect. This could be due to antialiasing setup in the drawing. 
a) can you tell me which tool you are using (mouse, touchscreen,... ?) 
b) can you redo your test set save the strokes and send the file :  this would help in analysis.

thx

**Comment by chrisj0192** [comment]
a) I'm using a mouse.
b) I redid a test. the stroke file is here: https://drive.google.com/drive/folders/1oxKOT3towY-lvWMM2xwiw3vT82setCEx?usp=sharing

**Comment by pubpub-zz** [comment]
I'm not able to reproduce your problem, but with the data you've sent I can see strange things : the first point is identical to the 3rd point when you are getting the problem, and strangely many times you have the same two first points. this is not systematic and in this case your strokes are similar to mine.
I've made a change to delete duplicated first points, can you try it ?
[ppInkArtefacts.zip](https://github.com/pubpub-zz/ppInk/files/6260957/ppInkArtefacts.zip)

**Comment by chrisj0192** [comment]
Hello friend, I tried your release and, in fact, it seems the problem was reduced, it's much better. But I've noticed that I can't make points now, I mean, I can't draw a point when I press left click (once).

Another issue is this one.
This is the link where I attached the screenshot:
https://drive.google.com/file/d/1sCC4M10P41R0avPWDKkdlUB6YzYKW_J7/view?usp=sharing (sorry for my spanish hehe)

"not enough memory resources are available to process this command." it's a strange issue but it is not the first time it appeared on my screen.

I'm sending you this observation because it's quite similar to another problem (I don't have a picture about it) when I draw a lot of things on my screen. :'( . In summary, I can tell that it appears when you make a lot of strokes on your screen and you might notice that the antialiasing starts to get unstable (especially when your new stroke is close to other strokes, appearing extra strokes) I'm not sure if it's a problem with my ram or something like that. that's why it's weird for me.

**Comment by pubpub-zz** [comment]
About the issue, I've detected the same issue and it is fixed keeping the same fix about your issue:
[ppInkAterfacts2.zip](https://github.com/pubpub-zz/ppInk/files/6271206/ppInkAterfacts2.zip)

About the crash, can you tell me : 
a) which version of Widows 10 you are running
b) what is your overall desktop surface in pixels (taking into acount all screens and there relative positions)
c) what are the actions you are doing before getting this issue ?
b) using the task manager / details tab, can you give me some info on the used memory, when starting ppInk, when starting inking, during drawings ( do you observe some memory increase, up to which amount, reductions( down to?), when the crask is reported?

finally a short question :  would you be interrested in improving the spanish translation ? 😊

**Comment by chrisj0192** [comment]
Hello friend. I hope you are well. I tested the new file and it's ok. Great! thanks!.

About the crash, these are my answers.
a) I'm using Microsft Windows 10 Home Single Language, Version: 20H2 (SO build 19042.867)
b) I'm not sure what you mean but my screen resolution is px.
c) Hmm when I teach math, the problem appears before I draw a lot of strokes on my screen (for example explaining an equation) or when I make a new stroke close to another stroke I made, the antialiasing starts getting unstable (extra lines start appearing). I have a short video about it. This is the link where I attach it: https://drive.google.com/drive/folders/1E8THR_DQHdtz95N9ruNPmPwX3cgaX-71?usp=sharing
d) I took some screenshots to show you some results :) I left the link in my answer C.

Finally, yes, I'm interested in contributing with the Spanish translation. What I need to do ? hehe. btw I don't know much about programming. I'd like to learn a little bit more.

**Comment by pubpub-zz** [comment]
a) about translation. I've prepared the file to modify
[es-latam.txt](https://github.com/pubpub-zz/ppInk/files/6288585/es-latam.txt)

Place it in lang folder
start editing it. You can view your updates just changing back and forth from/to spanish (file is reloaded everytime you select a new langage in options)

b) about your problems/crash : can you send the crash.txt file from ppink ; the dialog you've got indicates an exception fired during the exception processing. I hope to have more details in the file.
thanks

**Comment by pubpub-zz** [comment]
Hi @chrisj0192,
I've tried to redo the same amount of strokes and I'm getting a memory issue. I'm working on it. a first part of the fix is to move to 64bits application. attached:
[ppInk_x64.zip](https://github.com/pubpub-zz/ppInk/files/6303161/ppInk_x64.zip)

It should be better. can you confirm? Also, are you seeing any change on the occurence of the artifacts ? if getting some, can you change the priority of the process priority to high and check if there is any impact.

**Comment by chrisj0192** [comment]
Hello friend !, I hope you are well, these days I was pretty busy with my work that's why I couldn't reply you. I've tested the updated program and I've noticed that you solved the issue finally hehe. Thank you very much for that. I think you don't need my crash file anymore hehe.
Finally before closing this issue, I noticed a little issue with ellipse, rectangle and line drawings, mmm to be more specific when I draw a ellipse with alt + a key (customed) sometimes this ellipse dissapears (an ellipse that I've drawn) it's weird this issue because before it didn't happen. it happens with rectangle and line drawings too (but just sometimes). you could try it and you may notice it.

By the way, this is my contribution with the spanish translation. Here you are:
[es-latam.txt](https://github.com/pubpub-zz/ppInk/files/6341194/es-latam.txt)

If you need any other help, just let me know and I hope I can help you again hehe.

thanks in advance.

**Comment by pubpub-zz** [comment]
Great!
as said for this version  I had to do a lot of changes and Some regression may have appeared as some you have noticed.
About Ctrl+Alt+G, the problem has been reproduced and is under analysis.
About ellipse dissapearring, I have not been able to reproduce it. Can you try to make a video ?

Thanks for translation will commit your translation. for further updates request, can you subscribe to #17 ( I'm using this issue to raise alert about). Thanks 😊

**Comment by pubpub-zz** [comment]
Fix of Ctrl+Alt+G.
[ppInkx64ToggleFad.zip](https://github.com/pubpub-zz/ppInk/files/6346901/ppInkx64ToggleFad.zip)

**Comment by chrisj0192** [comment]
Hello friend! thank you so much for the updated program. Here you are a short video about ellipses, rectangles, lines and arrows disappearing. (it happens sometimes).
https://user-images.githubusercontent.com/20199672/115987335-3b097200-a57a-11eb-9b0a-f4aa27bf2f99.mp4

I hope you can fix that bug :)
Thanks  in advance!

**Comment by pubpub-zz** [comment]
Hi,
Thanks for the video. This could be linked with some issues with Magnetic effect : In order to help can you try to reproduce the issue as soon as you get the issue, change pen to hand writing, just draw one stroke and then save the strokes into a save.strokes.txt and send me the file.

Also, I increased my tests about your artefacts and I've finally been able to reproduce something which seems to be related. It occurs when you have lots of strokes, and *only* when you are drawing with the mouse : the problem does not appear when  using touchscreen for example.

I've make a new version using an intermediate InkCanvas to input data. Can you check if you can get artefacts. 
[ppInk64_25a.zip](https://github.com/pubpub-zz/ppInk/files/6372030/ppInk64_25a.zip)

note for this interim version, erasure swiping through strokes is not working.

**Comment by chrisj0192** [comment]
Hello, I'm attaching the stroke file. I did what you told me. Regarding ppInk64_25a.zip, I still have the same problem :(

[test 5.strokes.txt](https://github.com/pubpub-zz/ppInk/files/6372767/test.5.strokes.txt)

Thanks in advance.

**Comment by pubpub-zz** [comment]
when you say "I still have the same problem" Do you mean the strokes dissapearing or the artefacts?

**Comment by chrisj0192** [comment]
Yes, the strokes keep disappearing. I doubt the issue is related to my config because when I tested ppInk64_25a.zip (without changing anything) I started drawing ellipses and some ellipses disappeared automatically. It's weird because I didn't have this problem before the memory leak was fixed.

[config.zip](https://github.com/pubpub-zz/ppInk/files/6381285/config.zip)

About test 5.strokes.txt I confirm what you said. The 2nd ellipse disappeared, after that, I drew a stroke.

**Comment by pubpub-zz** [comment]
Still can not reproduce it 😭
Can you check if removing the magnetic, you are still able to reproduce the problem

**Comment by chrisj0192** [comment]
Yes, with or without magnetic effect my problem continues 😭 by the way, I was testing version 1.5 RC 2 and I don't have that problem. My issue may be related to memory leak fixed or duplicate points fixed, I'm not sure, I hope you can find it. I was thinking of closing this issue because it's related to the antialiasing issue (which is now resolved) and opening another issue, besides that I have some observations regarding saving and selecting objects (to give you some ideas for improvement). Ppink is much better now from my point of view :) I've not used zoomit again hehe.

**Comment by pubpub-zz** [comment]
I would rather keep this issue to deal with the memory leak and artefacts in drawing for better history tracking. for your improvement ideas, I agree about opening a new issue.
I've reviewed all commits since 1.5 RC2 and I can not identify an explanation. If you agree, I will reset my branch and reintroduce changes one step at a time and propose you the beta version for test to identify where the bug is introduced. OK for you?

**Comment by pubpub-zz** [comment]
This is the first version with spanish 😉, an issue fix at startup and the major memory leak 
[ppInkRebase1.zip](https://github.com/pubpub-zz/ppInk/files/6391473/ppInkRebase1.zip)
No drawing dissapearance ?

**Comment by chrisj0192** [comment]
Hello friend, Ok, so I leave this issue open and I'll open another one. Regarding ppInkRebase1.zip, I tested it and I can confirm that I had no problems with shapes (there's no drawing disappearance) :)

**Comment by pubpub-zz** [comment]
Great! next pack of commits

[ppInkRebase2.zip](https://github.com/pubpub-zz/ppInk/files/6405041/ppInkRebase2.zip)

Those are secondary fixes (including the first part of fix I've proposed for the artefacts and enhancements (includes SubToolBar)
Tell me If Shapes are still presents)
Tell me how it is Thanks again for your help

**Comment by chrisj0192** [comment]
Hello friend, yes I do confirm that in ppInkRebase2.zip I had problems with the forms (there are drawing disappearances) When I ran ppink (I pressed alt + ctrl + g) and drew my 2nd arrow, this arrow disappeared automatically. I tried drawing rectangles and ovals too and I had the same problem (shape disappearing), This happens only sometimes.

Just in case, I'm leaving my stroke files for you to review and hope they can help you find the problem.

[Test 6 (2nd arrow disappearance).strokes.txt](https://github.com/pubpub-zz/ppInk/files/6408569/Test.6.2nd.arrow.disappearance.strokes.txt)
[Test 7 (rectangle and oval disappearances).strokes.txt](https://github.com/pubpub-zz/ppInk/files/6408570/Test.7.rectangle.and.oval.disappearances.strokes.txt)

PD: I turned off the magnetic effect.

Thanks in advance for your attention.

**Comment by pubpub-zz** [comment]
I've reviewed the code and can not find any explanation... My current assumption is that the change to 64bits might be an explanation.
Ive generated a new version 32bits I've also activated a window to get the debug. When you are drawing an ellipse, a rect, or an arrow, a indication of the object will be written in the window. Can you tell me if when the shapes are dissapearing a line is reported.
[ppInk02a.zip](https://github.com/pubpub-zz/ppInk/files/6411800/ppInk02a.zip)

**Comment by pubpub-zz** [comment]
in addition to your feedbacks about [ppInk02a.zip](https://github.com/pubpub-zz/ppInk/files/6411800/ppInk02a.zip)
can you do also a test with this 64 bits version ( this version should normally have your bug):
[ppInk64_06a.zip](https://github.com/pubpub-zz/ppInk/files/6435002/ppInk64_06a.zip)

Can you start it, go into options, open debug window, and send me the content (select all and copy): It will list the loaded modules and path I would like to check the versions of DLLs you are using

Thanks

**Comment by pubpub-zz** [comment]
Hi @chrisj0192,
sorry to bother you but I foresee to release a new version soon and I would have liked to be able to close this issue before. Any results to share ?

**Comment by chrisj0192** [comment]
Hello friend! I apologize for not answering you earlier, I had a lot of work and was trying to deal with it. Regarding the versions I have to tell you that I always had the problem with 32bits and 64 bits. I did a lot of testing and noticed that with 32 bits the shape disappearing problem is much less than 64 bits. I redid the test with 1.5 RC2 version and I got the same problem of shape disappearing. I redid other tests with other versions 1.4, etc. and I got the same issue.

A short video with 32bit ppink.
https://user-images.githubusercontent.com/20199672/118589593-4a896e80-b766-11eb-9d59-612143931430.mp4

A short video with 64bit ppink.
https://user-images.githubusercontent.com/20199672/118589610-5117e600-b766-11eb-8065-6f937e6292ca.mp4

This is my DLL content that I copy from the prompt.
[DLLs.txt](https://github.com/pubpub-zz/ppInk/files/6498344/DLLs.txt)

That's all for now.
By the way, I saw a suggestion section. I was only going to create one but but there is already one hehe. I will comment some suggestions to improve ppink.

Thanks for your patience!

**Comment by pubpub-zz** [comment]
Thanks so much @chrisj0192 for you helpful results. I've progressed in the investigation.
I've prepared some new traces, can you do the same capture with this new version:
[ppInk64_18a.zip](https://github.com/pubpub-zz/ppInk/files/6503932/ppInk64_18a.zip)
thanks

**Comment by chrisj0192** [comment]
Hello my friend! Here's my test. I like doing this kind of work hehe.

This is a short video to show some  strokes and their results.
https://user-images.githubusercontent.com/20199672/118727002-b831ab00-b7f7-11eb-9d5e-3159096c5cf8.mp4

This is my DLL file (without results).
[DLLs 2.0.txt](https://github.com/pubpub-zz/ppInk/files/6504559/DLLs.2.0.txt)

This is my DLL file (including results).
[Results.txt](https://github.com/pubpub-zz/ppInk/files/6504560/Results.txt)

I'll stay tuned to your comments.

Thanks in advance.

**Comment by pubpub-zz** [comment]
Your test are great and your analysis is good!
It confirms what I would not have guess at the beginning with X0 being this long negative number.
I will prepare a new test version tomorrow for next evening 😪 😉
Just to confirm : you are using a mouse. Aren't-you? Can you tell which model it is and what are the driver / version ?

thx

**Comment by chrisj0192** [comment]
Yes, my mouse is Genius DX-110 Black USB Mouse (black color) and its driver version is 10.0.19041.1

**Comment by chrisj0192** [comment]
I'm trying to do my best hehe thanks a lot.
Yes, my mouse is Genius DX-110 Black USB Mouse (black color) and its driver version is 10.0.19041.1

**Comment by chrisj0192** [comment]
I'm trying to do my best hehe thanks a lot.

About my mouse is Genius DX-110 Black USB Mouse and its driver version is 10.0.19041.1

**Comment by pubpub-zz** [comment]
I've got it!!!!!!
There is 4 events : CursorDown (en starting to draw a stroke), MouseDown( when a mouse button down is generated by mouse, pen or touchscreen), Stroke (wne the stroke is finished) and MouseUp (when the "button" is released and during this event, I reset X0) During all my test, Events were raised in the order CursorDown/MouseDown/Stroke/MouseUp but on your computer, you got sometimes MouseUp before Stroke: I was not aware of this asynchronous order.
I've modified the code to cope with this:
[ppInk64_19a.zip](https://github.com/pubpub-zz/ppInk/files/6511731/ppInk64_19a.zip)

Can you confirm the problem went away. If you see something odd, can you redo a record with the debug window
Thanks!

**Comment by chrisj0192** [comment]
Hello friend! let me tell you that you did it again! hehe, Congratulations. You're amazing. The problem was finally solved. I have no observation and I haven't seen anything odd for now, only this one (I'm attaching a screenshot below).

![1](https://user-images.githubusercontent.com/20199672/118903331-33b95800-b8dd-11eb-9869-5a5fb9ad44d1.png)
The "enable all following hotkeys..."  option is on the eraser option (I can't see eraser option correctly and it's a bit difficult to set it with a key). May you adjust its position?

I think this is all for now.  I'll close this issue because it was solved.

Let me tell you

**Comment by chrisj0192** [comment]
I found some bugs or well, I'm not sure if they are bugs, but I'll report them in the other issue called "ideas for improvements"  because they are related to the snapshot, zoom (frozen area) options.  Another is that it seems that the color menu is activated automatically when I select my highlighter with a custom key (it hardly happens). I'll have to try ppink for longer to see If those are really bugs. For now I'm pretty grateful for your help in solving "the shape disappearance". 

Thank you so much!

**Comment by pubpub-zz** [comment]
If you don't mind please keep bugs in this issue : I've not worked yet on the improvements and would rather keep tracks of bugs on beta versions in the same issue where the beta has been proposed

**Comment by chrisj0192** [comment]
Hello friend! I hope you are ok ... these days I've been testing the ppInk64_19a.zip and I've noticed that there are some bugs but I'm not sure if they are bugs. I recorded some videos to try to explain what I found. Hmm, I'm not sure where I should post those observations, in this issue or in another issue you created called "looking for beta tester for ...". I'd like to share my observations, but I think this issue should be close because I consider that memory leak and shape disappearance were solved up to 90% as far as I know hehe. I have to keep trying to confirm that it was fully resolved but for now I have no problem with it. Also those new observations that I'll show you aren't related to this issue.

I'll wait for your answer.

Thanks.

PD: I'm testing ppInk64_24a.zip.

**Comment by pubpub-zz** [comment]
Hi  @chrisj0192,
feel free to report your observations in this issue :  quite often, they are related to changes introduced in the beta sent through this chat and that way it is easier for me to track. If the problems are to be reported out in an new issue, I will create a new issue. Looking forward to your inputs, I'm finalising a new official release and will try to fix them before.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz this is my report about what I found in ppInk64_19a.zip (at that moment). I've noticed that some bugs were fixed in ppInk64_24a.zip (pre-release of 1.6 RC1).

Anyway I'm attaching the videos that I have and I'll explain each one of them.

- About the clipboard:
In this video I'm trying to explain two things.
1. I'd not realized that when I took a screenshot with ppink, this screnshot was saved in ppink folder and at the same time this capture was in the clipboard. that was incredible for me. But sometimes I didn't want to save a screenshot in my ppink folder so, I was thinking if you could consider implment another funcition similar to snapshot but just to copy a draw to the clipboard (without disappearing the drawing) Because por example sometimes I closed ppink by accident and I lost all what I did and that was frustrared. With clipboard I make sure that "for example press "ctrl + c" to copy my drawing to clipboard (without saving the screenshot).
Now, I don't know if there is an option to do that. if there is an option like that, please tell me to do it. 
OBS: the current snapshot is great! please dont modify it . I just try to explain an similar idea of the snapshot in order to have two ways to keep our drawing.
Snapshot is a great option and I like it, but if you could consider an option just only to copy to clipboard what we draw on the screen (without closing the drawing) , it would be nice. If you know and option to do that, please tell me,

2. Another one is about a bug that I found. I think in ppInk64_24a.zip (pre-release of 1.6 RC1) was fixed.
Anyway I'll explain you. this bug was about that when I took a screenshot with ppink ("ctrl + s" are my custom keys)  and then when I activated ppink again, when I press alt + A to start making an arrow, automatically appeared and arrow on the screen (its size was similat to the screenshot that I took) that issue can be seen in the video. (but I think this issue was solved in the new release as I told you, because I've not noticed again so far)

This is the video:
https://user-images.githubusercontent.com/20199672/120062747-40e6ed00-c029-11eb-822d-429d6d77c796.mp4

- About the highlighter pen test (pressing and holding a key)
In this issue I've noticed that when I run ppink (activated) when I pressed "alt + space" to start using my nightlighter (pressing thw two keys at the same time) then when I released those keys, the color pen/edit appeared automatically. that was weird.

Here you are a video what I'm trying to explain you:
https://user-images.githubusercontent.com/20199672/120063286-0599ed80-c02c-11eb-8980-11b6ca13e087.mp4

Besides, in ppInk64_24a.zip (pre-release of 1.6 RC1) when I run ppink for the first time and I press "alt + g", the color pen/edit appeared automatically. I've noticed the same issue but with "alt + b" (for example). But it appeared When I run the program for the first time (I mean when I run ppink for the first time)

- About the pen width:
In this video I've explained about this issue and I don't know if it's a bug but I want to show you anyway.
https://user-images.githubusercontent.com/20199672/120063571-5d852400-c02d-11eb-82cd-98e40ede9f96.mp4

That's all what I can explain about what I found.
Again I apologize if my post is quite long but I needed to explain you what I found hehe,

I'll be waiting for your answer.

Thanks in advance.

**Comment by pubpub-zz** [comment]
>     * About the clipboard:
>Because for example sometimes I closed ppink by accident and I lost all what I did and that was frustrared. 

I share your frustration and there is already a work around : when inking is stopped (through snapshot or standard exiting), your strokes are automatically saved and you can reload them restarting inking and then pressing the load button.

Also, when you long press snapshot ( or use alt+ Hotkey_snapshot) you can do a capture without exiting drawing.

> With clipboard I make sure that "for example press "ctrl + c" to copy my drawing to clipboard (without saving the screenshot).

if you don't want to keep images, just delete the files periodically. Not convinced it worth time and complexification.

>        OBS: the current snapshot is great! please dont modify it . I just try to explain an similar idea of the snapshot in order to have two ways to keep our drawing.
>        Snapshot is a great option and I like it, but if you could consider an option just only to copy to clipboard what we draw on the screen (without closing the drawing) , it would be nice. If you know and option to do that, please tell me,

when you long press snapshot button( or use alt+ Hotkey_snapshot) you can do a capture without exiting drawing.
I agree that something that could be interresting would be to have an option to swap way sto do a snapshot with and without continue inking.

>     2. Another one is about a bug that I found. I think in ppInk64_24a.zip (pre-release of 1.6 RC1) was fixed.
>        Anyway I'll explain you. this bug was about that when I took a screenshot with ppink ("ctrl + s" are my custom keys)  and then when I activated ppink again, when I press alt + A to start making an arrow, automatically appeared and arrow on the screen (its size was similat to the screenshot that I took) that issue can be seen in the video. (but I think this issue was solved in the new release as I told you, because I've not noticed again so far)

Yeap, found and fixed (you can try latest version from #98.

> This is the video:
> https://user-images.githubusercontent.com/20199672/120062747-40e6ed00-c029-11eb-822d-429d6d77c796.mp4
> 
>     * About the highlighter pen test (pressing and holding a key)
>       In this issue I've noticed that when I run ppink (activated) when I pressed "alt + space" to start using my nightlighter (pressing thw two keys at the same time) then when I released those keys, the color pen/edit appeared automatically. that was weird.

in the default configuration attached to delivery, Alt is used as temporary key:
![image](https://user-images.githubusercontent.com/4083478/120069968-ed0bf080-c088-11eb-955b-e0980eeabb6f.png)

If you want to have your hotkeys to properly process alt in shortcuts, uncheck it. Personnally I prefer to use Ctrl and/or shift and keep alt as temporary

> Here you are a video what I'm trying to explain you:
> https://user-images.githubusercontent.com/20199672/120063286-0599ed80-c02c-11eb-8980-11b6ca13e087.mp4

I'm confused about pen definition dialog box opening. can you provide more details. You can use CARNAC to show keyboards

> 
> Besides, in ppInk64_24a.zip (pre-release of 1.6 RC1) when I run ppink for the first time and I press "alt + g", the color pen/edit appeared automatically. I've noticed the same issue but with "alt + b" (for example). But it appeared When I run the program for the first time (I mean when I run ppink for the first time)

Not reproduced neither  : can you remind me what functions are behind alt+g and alt+b for you ?

> 
>     * About the pen width:
>       In this video I've explained about this issue and I don't know if it's a bug but I want to show you anyway.
>       https://user-images.githubusercontent.com/20199672/120063571-5d852400-c02d-11eb-82cd-98e40ede9f96.mp4

The first time I've coded that way and I wonder What is the best solution between the two options : like currently coded, you can recover the width before selecting the new pen. in another hand, It can be a little confiousing to not restart fro the current size. Can you open a new issue about this specific point opening the question to get other opinions.

> That's all what I can explain about what I found.
> Again I apologize if my post is quite long but I needed to explain you what I found hehe,

No. I appreciate the details and the videos in order to have a better understanding of the test case.

**Comment by pubpub-zz** [comment]
I propose to close the thread now

**Comment by chrisj0192** [comment]
Hello @pubpub-zz FIrst of all thank you so much for your answers.

>I share your frustration and there is already a work around : when inking is stopped (through snapshot or standard exiting), your strokes are automatically saved and you can reload them restarting inking and then pressing the load button.

I started using more available options of ppInk because I was doing the Spanish translation (I completed it about 98%, you might want to check it into #17) and sincerely I didn't know that there was a load button to load strokes that are automatically saved!!! I'm sorry for not noticing that but I don't know if you noticed that I don't use the toolbar hehe, I use ppInk with full hotkeys and mouse/touchpad and I hope I can use ppInk with a graphic tablet too.

So, Could I request you something? Could you please implement a hotkey for that load button?, because I'm less dynamic using the toolbar than using hotkeys and have also gotten used to using ppInk with keyboard and mouse. Please, consider it!
There are two options in load button, maybe it would be with a secondary toolbar? the same request for save button (with hotkeys) I'd would really like to configure those buttons with my own keys.

>Also, when you long press snapshot ( or use alt+ Hotkey_snapshot) you can do a capture without exiting drawing.

The same request for this one please, could you please implement a hotkey to custom this option (I'd like to set it with my own key too)

>if you don't want to keep images, just delete the files periodically. Not convinced it worth time and complexification.

Agree with you!

>when you long press snapshot button( or use alt+ Hotkey_snapshot) you can do a capture without exiting drawing.
I agree that something that could be interresting would be to have an option to swap way sto do a snapshot with and without continue inking.

I totally agree! as I told you above, I don't know If you could implement a hotkey for those two options because I'd like to set my own keys. I think that it would be more flexible. Maybe, with a secondary toolbar too? because there are two ways to take snapshots.

> If you want to have your hotkeys to properly process alt in shortcuts, uncheck it. Personnally I prefer to use Ctrl and/or shift and keep alt as temporary

I'm afraid that I've unchecked it some months ago, because I get more posibilities to combine more keys for available options. But, the main reason that I unchecked it and I love using Alt key is because when I combine alt + "a custom key" I can return to the pen! and that's amazing for me, that's the function the I love the most in ppInk. I think I commented you an example in dashed lines issue (for example, when I draw a rectangle holding alt + "my own configured hotkey" and release alt, the pen returns!. For me that's the best function because from my point of view, that function makes it quite dynamic with the keyboard! I can press and hold down alt and swap from one key to another key to make many shapes easily and fast (rectangles, arrows, ovals, lines including If I press twice, thrice or four times the same key to change its function and draw a colored rectangle, a colored oval, a black rectangle, etc).
I cannot do the same action with the shift and ctrl keys 😞.

>I'm confused about pen definition dialog box opening. can you provide more details. You can use CARNAC to show keyboards.

It happens if you press a color hotkey for a long time and if you set the "space" key for a color and you press and hold down that key, the color/pen edit window will appear and might disappear automatically.

> Not reproduced neither : can you remind me what functions are behind alt+g and alt+b for you ?
I explained it above.

>The first time I've coded that way and I wonder What is the best solution between the two options : like currently coded, you can recover the width before selecting the new pen. in another hand, It can be a little confiousing to not restart fro the current size. Can you open a new issue about this specific point opening the question to get other opinions.

For me, it's really confusing because I use the arrow and with it I cannot see how wide my pen is or which my last width was before swaping to another pen color (because when I swap from a pen to another one, the pen width restarts from the preset value (that's great). I think if it restarts when using the mouse wheel or hotkeys starting the pen width from its preset value, it will be more intuitive. Even so, I'm willing to open an issue about this topic. It would be interesting to get other opinions.

That's all what I wanted to tell you my friend in this issue my friend. I'll close this issue because as I told you days ago, this issue was fixed!

Thank you very much for your help, for your attention, for your support and don't hesitate to count on me!

**Comment by pubpub-zz** [comment]
> I started using more available options of ppInk because I was doing the Spanish translation (I completed it about 98%, you might want to check it into #17) and sincerely I didn't know that there was a load button to load strokes that are automatically saved!!! I'm sorry for not noticing that but I don't know if you noticed that I don't use the toolbar hehe, I use ppInk with full hotkeys and mouse/touchpad and I hope I can use ppInk with a graphic tablet too.

So do I :  for Graphic tablet it is just a question of money to buy it ... 🙄
About your way of working, have you try with 2 screens and leave the toolbar on secondary screen.

> So, Could I request you something? Could you please implement a hotkey for that load button?, because I'm less dynamic using the toolbar than using hotkeys and have also gotten used to using ppInk with keyboard and mouse. Please, consider it!

I thought it was already implemented.... Will put that in the stack... 👍 

> There are two options in load button, maybe it would be with a secondary toolbar? the same request for save button (with hotkeys) I'd would really like to configure those buttons with my own keys.

This is a little more tough : the button reacts immediately when loading default file and I would not like to slow down this use case to leave some time before loading file. I'm open to proposals

> I'm afraid that I've unchecked it some months ago, because I get more posibilities to combine more keys for available options. But, the main reason that I unchecked it and I love using Alt key is because when I combine alt + "a custom key" I can return to the pen! and that's amazing for me, that's the function the I love the most in ppInk. I think I commented you an example in dashed lines issue (for example, when I draw a rectangle holding alt + "my own configured hotkey" and release alt, the pen returns!. For me that's the best function because from my point of view, that function makes it quite dynamic with the keyboard! I can press and hold down alt and swap from one key to another key to make many shapes easily and fast (rectangles, arrows, ovals, lines including If I press twice, thrice or four times the same key to change its function and draw a colored rectangle, a colored oval, a black rectangle, etc).
> I cannot do the same action with the shift and ctrl keys 😞.

I think I have not been clear about about the alt key as temporary option: witht Alt as temporary option set, if you set Ctrl+R for rectangle, when you press Ctrl+R you will select Rectangle "definitely" but if you press Ctrl+Alt+R, you will activate Rectangle and this will be removed immediately after releasing Alt.  If you set the hotkey to Ctrl+Alt+R, you will get exactly the same behavior.

> > I'm confused about pen definition dialog box opening. can you provide more details. You can use CARNAC to show keyboards.
> It happens if you press a color hotkey for a long time and if you set the "space" key for a color and you press and hold down that key, the color/pen edit window will appear and might disappear automatically.

Thanks  Will try it

> > Not reproduced neither : can you remind me what functions are behind alt+g and alt+b for you ?
> 

Can you send me your hotkeys.ini and config.ini files please, It should ease my debug

> 
> > The first time I've coded that way and I wonder What is the best solution between the two options : like currently coded, you can recover the width before selecting the new pen. in another hand, It can be a little confiousing to not restart fro the current size. Can you open a new issue about this specific point opening the question to get other opinions.
 
> For me, it's really confusing because I use the arrow and with it I cannot see how wide my pen is or which my last width was before swaping to another pen color (because when I swap from a pen to another one, the pen width restarts from the preset value (that's great). I think if it restarts when using the mouse wheel or hotkeys starting the pen width from its preset value, it will be more intuitive. Even so, I'm willing to open an issue about this topic. It would be interesting to get other opinions.

I seen what you mean : will put in the stack 👍 


Note : I you can please answer the few questions I've raised in this thread : even closed you can carry on chating in.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz

> This is a little more tough : the button reacts immediately when loading default file and I would not like to slow down this use case to leave some time before loading file. I'm open to proposals.

For now, a hotkey is enough for me to load the last strokes that I made if I accidentally close ppInk. Don't worry about it friend.
In addition I commented more details about it in #106 

>I think I have not been clear about about the alt key as temporary option: witht Alt as temporary option set, if you set Ctrl+R for rectangle, when you press Ctrl+R you will select Rectangle "definitely" but if you press Ctrl+Alt+R, you will activate Rectangle and this will be removed immediately after releasing Alt. If you set the hotkey to Ctrl+Alt+R, you will get exactly the same behavior.

Yes, I understood what you said. I was trying those combinations with alt key but it's a bit hard to use three keys at the same time especially combining them and I feel like my left little finger gets tired much faster due to pressure and I need to be more focused when using the keys and graphing at the same time. Although for now the "alt" key is enough for me to make many combinations according to my work.

>Thanks Will try it

Don't forget to try it!

>Can you send me your hotkeys.ini and config.ini files please, It should ease my debug

Yes, here you are:
[config & hotkeys.zip](https://github.com/pubpub-zz/ppInk/files/6606211/config.hotkeys.zip)

>I seen what you mean : will put in the stack
I gave my report about it in #107

I'll stay tuned to your comments.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz

> This is a little more tough : the button reacts immediately when loading default file and I would not like to slow down this use case to leave some time before loading file. I'm open to proposals.

For now, a hotkey is enough for me to load the last strokes that I made if I accidentally close ppInk. Don't worry about it friend.
In addition I commented more details about it in #106 

>I think I have not been clear about about the alt key as temporary option: witht Alt as temporary option set, if you set Ctrl+R for rectangle, when you press Ctrl+R you will select Rectangle "definitely" but if you press Ctrl+Alt+R, you will activate Rectangle and this will be removed immediately after releasing Alt. If you set the hotkey to Ctrl+Alt+R, you will get exactly the same behavior.

Yes, I understood what you said. I was trying those combinations with alt key but it's a bit hard to use three keys at the same time especially combining them and I feel like my left little finger gets tired much faster due to pressure and I need to be more focused when using the keys and graphing at the same time. Although for now the "alt" key is enough for me to make many combinations according to my work.

>Thanks Will try it

Don't forget to try it!

>Can you send me your hotkeys.ini and config.ini files please, It should ease my debug

Yes, here you are:
[config & hotkeys.zip](https://github.com/pubpub-zz/ppInk/files/6606211/config.hotkeys.zip)

>I seen what you mean : will put in the stack

I gave my report about it in #107

I'll stay tuned to your comments.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz

> This is a little more tough : the button reacts immediately when loading default file and I would not like to slow down this use case to leave some time before loading file. I'm open to proposals.

For now, a hotkey is enough for me to load the last strokes that I made if I accidentally close ppInk. Don't worry about it friend.
In addition I commented more details about it in #106 

>I think I have not been clear about about the alt key as temporary option: witht Alt as temporary option set, if you set Ctrl+R for rectangle, when you press Ctrl+R you will select Rectangle "definitely" but if you press Ctrl+Alt+R, you will activate Rectangle and this will be removed immediately after releasing Alt. If you set the hotkey to Ctrl+Alt+R, you will get exactly the same behavior.

Yes, I understood what you said. I was trying those combinations with alt key but it's a bit hard to use three keys at the same time especially combining them and I feel like my left little finger gets tired much faster due to pressure and I need to be more focused when using the keys and graphing at the same time. Although for now the "alt" key is enough for me to make many combinations according to my work.

>Thanks Will try it

Don't forget to try it!

>Can you send me your hotkeys.ini and config.ini files please, It should ease my debug

Yes, here you are:
[config_&_hotkeys.zip](https://github.com/pubpub-zz/ppInk/files/6606224/config_._hotkeys.zip)

>I seen what you mean : will put in the stack

I gave my report about it in #107

I'll stay tuned to your comments.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz

> This is a little more tough : the button reacts immediately when loading default file and I would not like to slow down this use case to leave some time before loading file. I'm open to proposals.

For now, a hotkey is enough for me to load the last strokes that I made if I accidentally close ppInk. Don't worry about it friend.
In addition I commented more details about it in #106 

>I think I have not been clear about about the alt key as temporary option: witht Alt as temporary option set, if you set Ctrl+R for rectangle, when you press Ctrl+R you will select Rectangle "definitely" but if you press Ctrl+Alt+R, you will activate Rectangle and this will be removed immediately after releasing Alt. If you set the hotkey to Ctrl+Alt+R, you will get exactly the same behavior.

Yes, I understood what you said. I was trying those combinations with alt key but it's a bit hard to use three keys at the same time especially combining them and I feel like my left little finger gets tired much faster due to pressure and I need to be more focused when using the keys and graphing at the same time. Although for now the "alt" key is enough for me to make many combinations according to my work.

>Thanks Will try it

Don't forget to try it!

>Can you send me your hotkeys.ini and config.ini files please, It should ease my debug

Yes, here you are:
[config_&_hotkeys.zip](https://github.com/pubpub-zz/ppInk/files/6606224/config_._hotkeys.zip)

>I seen what you mean : will put in the stack

I gave my report about it in #107

I'll stay tuned to your comments.

Thanks!

**Comment by pubpub-zz** [comment]
> > Thanks Will try it
> 
> Don't forget to try it!

Try the latest version, I have not been able to reproduce but I found something not so clear and cleaned it.

> PS: I've noticed an issue regarding clipArts and have a suggestion as well. I don't know whether to comment on it here or create a new issue.
> 

Complete existing one is ok for me

**Comment by chrisj0192** [comment]
Hello @pubpub-zz I know this issue was closed, but last time I told you that I wanted to show you something related to clipArt and I'll show you here in this issue.

I noticed that when loading a new picture in the clipArt box, it's ok. when selecting that new picture to set as a clipArt1 to use with a shortcut, it's ok too. when saving to files and I start using it, no problem, but when pressing exit button of ppInk and I open ppInk I find that my new picture is duplicated. is it normal? I noticed that it happens with clipArt1/2/3.
https://user-images.githubusercontent.com/20199672/121767200-7f49d500-cb1c-11eb-8e21-939c136689c5.mp4

Another one is a suggestion If you might want to consider it (regarding clipArt)
I suggest that a selected picture be put on the screen only once and then ppInk automatically returns to the pen. In this case I think it could be managed using a check button to select between those two options (put a picture on the screen once or put a picture on the screen many times). From my point of view it would be more intuitive and dynamic if a picture is put on the screen only once when using many pictures at the same time, for example when teaching (If we forget to swap to the pen).
https://user-images.githubusercontent.com/20199672/121767527-32ff9480-cb1e-11eb-9f42-c42b04527e30.mp4

Another issue that I found is about the toolbar.
I think the video shows it clearly.
Here you are:
https://user-images.githubusercontent.com/20199672/121767588-a2758400-cb1e-11eb-98cb-e43ef4959dfd.mp4

Those are all my observations.

Thanks in advance for your attention.

**Comment by pubpub-zz** [comment]
> I noticed that when loading a new picture in the clipArt box, it's ok. when selecting that new picture to set as a clipArt1 to use with a shortcut, it's ok too. when saving to files and I start using it, no problem, but when pressing exit button of ppInk and I open ppInk I find that my new picture is duplicated. is it normal? I noticed that it happens with clipArt1/2/3.

Not normal thanks for report. Found and fixed in beta below

> Another one is a suggestion If you might want to consider it (regarding clipArt)
> I suggest that a selected picture be put on the screen only once and then ppInk automatically returns to the pen. In this case I think it could be managed using a checkbox to select between those two options (a picture on the screen only once [activated] or a picture loop on the screen [deactivated]). From my point of view it would be more intuitive and dynamic if a picture is put on the screen only once when using many pictures at the same time, for example when teaching (If we forget to return to the pen). What do you think about it?

What you are asking is already covered with the alt as temporary command. about this I found from your videos that there was a bug : with the Alt as temporary option off you should not return back to your previous tools when releasing... fixed in beta release attached
From my opinion you should keep the alt key for temporary commands (you should be able to replace all your alt+ by shift+...)

> Another issue that I found is about the toolbar icons, they aren't displayed when activating them again in their checkboxes.
> I think the video shows it clearly.

I had this issue fixed with most of the buttons but  I forgot to check the existing ones from gInk... fixed now:

The beta release for test 😉
[ppInk64_varFix12a.zip](https://github.com/pubpub-zz/ppInk/files/6642110/ppInk64_varFix12a.zip)

**Comment by chrisj0192** [comment]
Hello @pubpub-zz , I wanted to test it but ppInk64_varFix12a.zip doesn't open by clicking on the main icon of ppInk and the blue tray icon doesn't show up.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz I finally tested the last beta release. and in short, these are my results:
- About the toolbar icons when activating /deactivating icons through their checkbox works perfectly now. They are displayed properly.
- About picture duplication was also fixed. I have no problem with it.
- About the alt + shortcut, Here I have many problems, my questions is: have you change the function of "alt" key? because when I activate it through "process Alt as temporary command..." it works. But when I deactivate that option. It doesn't work properly and It's really confusing because I have some keys with only "letter" and others with "alt + the same letter" causing different and mixing results as expected. I use Shift/Ctrl  keys for other purposes for example to take a screenshot (snapshot), load/save strokes, zoom, etc. But I need to use alt + shortcut because when I teach my students, it allows me to be more fluent in my move, and dynamic swaping one function to other functions  saving a lot of time. I can't get the same result by pressing three keys at the same time ctrl/shift + alt + shortcut, dealing with it is harder for me. that's the resaon by I use "alt" for my principal activities (shapes/pictures) and I deactivate "process alt as ....". I don't know if it's a new bug or it was a change.

I'll stay tuned for your answer.

Thanks.

**Comment by chrisj0192** [comment]
>What you are asking is already covered with the alt as temporary command. about this I found from your videos that there was a bug : with the Alt as temporary option off you should not return back to your previous tools when releasing... fixed in beta release attached
From my opinion you should keep the alt key for temporary commands (you should be able to replace all your alt+ by shift+...)

I forgot to read this comment ... omg, well.... honestly, I never thought that "that bug" was a bug, well....for me it's been the best bug that I found in your program for reasons I explained above. 
I don't want to lie to you, but it discourages me to know that.
That bug allows me to use alt +Q to draw rectangle and then releasing and use only Q to use black pen (an example of many eamples). I decided to work with this configuration with some keys because we use our left hand to use keys and our right hand to handle and use the mouse. In my case, I set many letter keys to use color pens and they are near my left hand. I cannot get the same results with shift (because it seems to have a temporary magnetic effect when combined with a shape for example a line)
I'd like to tell you something:
Months ago I was looking for a program that be used with shortcut keys and I found ZoomIt, Idk If you know that annotation tool. but its configuration was wonderful for me. You might want to try it if you don't know it. Let me tell you that ZoomIt shortcuts and that alt bug that you fixed in ppInk are similar and that's why I've been usig those combinations that  you watched in my test videos. 
I tried EpicPen (I honestly don't like its usability because one have always use shift key if you do not want to get a mixed result with a windows key combination that already has a function), Pointofix (here you can use alt to swap from one function to another but lacks width selection), Pointer Focus (not very intuitive and lacks many tools) and I've used other annotation tools that honestly I never felt comfortable with its features. Despite all that. I keep using ZoomIt because I liked its configuration with the keys, quite easy to use, despite the few features it has. For me ZoomIt was the best annotation tool I found due to usability with keys, the only thing I didn't like was that I can't customize colors. Then I found gInk that was a wonderful tool at that moment but it didn't have shapes (line/oval, etc), but finally I found ppInk and fell in love with "that bug" that you've fixed now 😭
I'm sorry my friend but I think I told you in another comment that "I deactivated "process Alt ..." option because I could combine a specific letter and alt + that specific letter (the same letter key) to get more possibilities to combine keys + releasing and returning to the pen automatically (similar to ZoomIt). I don't like to compare, but really I found that former feature in ppInk (similar function of how ZoomIt works with the keys) quite great. don't bother with me....

These days I was testing your program to help you find bugs and for you to fix them but not to fix that "beautiful bug" that was very useful for me.😭
I hope you can understand me. If you have another solution, I'll really appreciate it. If not, don't worry I keep using ppInk with the last beta release about arrow proposal because as I told you, your program is the best annotation tool that I've found.

btw here you are some corrections about en-us lang that I found.
[en-us.txt](https://github.com/pubpub-zz/ppInk/files/6643432/en-us.txt)
Some lines that I corrected:
ListFillingsText = "No frame;With frame;Color filled;White filled;Black filled"
PanSubToolsHints = "Move one stroke\nCopy one stroke\nPan all strokes"
And few other words.

These are my latest version of the Spanish translation (99%):
[es-es.txt](https://github.com/pubpub-zz/ppInk/files/6643437/es-es.txt)
[es-latam.txt](https://github.com/pubpub-zz/ppInk/files/6643438/es-latam.txt)
To be honest, there's not much difference with each other. You could replace es-es in es-latam. (es-es file is standard Spanish, I mean words used in Spain and Latinamerican, there's not much difference).

And this is my proposal in English (I did few changes)
[en-us (proposal).txt](https://github.com/pubpub-zz/ppInk/files/6643446/en-us.proposal.txt)
you might want to evaluate it. it's up to you.

**Comment by pubpub-zz** [comment]
@chrisj0192,

>     * About the alt + shortcut, Here I have many problems, my questions is: have you changed the function of "alt" key? because when I activate it through "process Alt as temporary command..." it works. But when I deactivate that option. It doesn't work properly and It's really confusing because I have some hotkeys with only "letter" and others with "alt + the same letter" causing different and mixed results as not expected. I use Shift/Ctrl  keys for other purposes for example to take a screenshot (snapshot), load/save strokes, zoom, etc, I use letter keys for each pen color. But I need to use alt + shortcut because when I teach my students, it allows me to be more fluent in my move, and dynamism swaping one function to other functions  returning  to the pen automatically, saving a lot of time. I can't get the same result by pressing three keys at the same time ctrl/shift + alt + shortcut, dealing with it is harder for me. that's the reason why I used "alt" for my principal activities (shapes/clipArts) and that's the reason why I deactivated "process alt as ....". I don't know if it's a new bug or it was a change. 😭

On the current beta version, The behavior expected is :
**with "Alt As Temporary" disabled **: Q and alt+Q are differentiated but the tool/pen behind alt+Q is kept active all time long
Can you confirm this is the result of your test.

Based on your hotkeys.ini I've drawn the following keyboard layout:
![image](https://user-images.githubusercontent.com/4083478/121800474-67647500-cc32-11eb-97de-739f66cf0a64.png)

Inhere, Green identifies Shift+, Pink identifies  Ctrl+, Yellow identifies alt+, and blue for unmodified keys
When you look,only 4 keys will interfering with alt as temporary option: What about remapping Q to Tab (for this version you will have to directly modify hotkeys.ini), and RBG to 5,6,7 
could this reallocation match your way to use keyboards?

For another solution I will see what I can do

Also about translation. Thanks, but can you publish them in #17 thread for traceability. Thanks

**Comment by chrisj0192** [comment]
> On the current beta version, The behavior expected is :
**with "Alt As Temporary" disabled **: Q and alt+Q are differentiated but the tool/pen behind alt+Q is kept active all time long
Can you confirm this is the result of your test.

Yes, deactivating "process alt" alt +Q and Q are differentiated but how I've explained above  the problem is when I draw rectangles using  alt + Q and it gets stuck and doesn't return to the pen automatically (this one is really annoying for me) (I know that "releasing" is a function of that process alt option", well I have no more words. I've tried to explain my point of view with details above.
And also my hotkey configuration was changed due to the lasso and I was trying to get other combinations to be more comfortable. on the other hand I don't change colors with numbers because I use the keyboard without seeing it. it's easy for me managed colors with letters. I have gotten used to that configuration since using ZoomIt (around a year and a half). And it's practical for me, it's not easy to press shirt + O or shift + L with one hand and it's not easy to adapt to a new configuration in no time knowing that tomorrow I have to teach my students using a new configuration. Maybe if you see me how it use your program with my hands you might understand much better. No more to say and don't worry I understand your approach and as I told you I'll keep using with the last beta version (arrow proposal). I like helping people but sincerely I don't like to bother you with my requests. it's uncomfortable for me to do that.

Thanks you anyway 😊

**Comment by pubpub-zz** [comment]
@chrisj0192 
    I left for so long time a bug in the system it became a feature... I found a way to keep compatibility

[ppInk64_f4C13a.zip](https://github.com/pubpub-zz/ppInk/files/6643713/ppInk64_f4C13a.zip)

In this version, right-click on Alt as Temporary and you will get the interim state which should correspond to the "buggy" behavior

the expected behavior should be for me;
- UnChecked = Alt+R and R will NOT be distinguished ; if alt +Shortcut  is depressed, tool and/or pen are active until alt release
- (new) Indetermined(obtainted by right click) = Alt+R and R will be distinguished ; ONLY Shortcuts using Alt will be activated until alt released and others will be 'permanent'
- UnChecked  = Alt+R and R will be distinguished ; ALL Shortcuts EVEN using Alt will be activated permanently

Can you confirm all those three behaviors.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz
First of all, I want to thank you for implementing this pseudo-bug in ppInk 😊. This function makes me feel quite happy. I say pseudo-bug because it was never a bug for me hehe. In fact, I consider that "bug" as an intermediate state of process 'alt' enabled and process 'alt disabled or well a mix of those two.
thank you sincerely.

I was testing this feature and everything is ok. There is no problem with the 3 behaviors you mentioned.
This feature is quite special for me. It's my main favorite feature.

Thanks again and if I notice anything odd, don't worry, I'll let you know. For now everything is ok.
